Mathlib/Algebra/Homology/TotalComplexShift.lean

```diff
@@ -0,0 +1,206 @@
+    dsimp at h hpq
+    dsimp [totalShift₁XIso]
+    rw [ι_D₁_assoc, Linear.comp_units_smul, ι_totalDesc_assoc, ι_D₁,
+      ((shiftFunctor₁ C x).obj K).d₁_eq _ rfl _ _ (by dsimp; omega),
+      K.d₁_eq _ (show p + x + 1 = p + 1 + x by omega) _ _ (by dsimp; omega)]
+    dsimp
+    rw [one_smul, assoc, ι_totalDesc, one_smul, Linear.units_smul_comp]
+  · rw [D₁_shape _ _ _ _ h, zero_comp, D₁_shape, comp_zero, smul_zero]
+    intro h'
+    apply h
+    dsimp at h' ⊢
+    omega
+
+lemma D₂_totalShift₁XIso_hom (n₀ n₁ n₀' n₁' : ℤ) (h₀ : n₀ + x = n₀') (h₁ : n₁ + x = n₁') :
+    ((shiftFunctor₁ C x).obj K).D₂ (up ℤ) n₀ n₁ ≫ (K.totalShift₁XIso x n₁ n₁' h₁).hom =
+      x.negOnePow • ((K.totalShift₁XIso x n₀ n₀' h₀).hom ≫ K.D₂ (up ℤ) n₀' n₁') := by
+  by_cases h : (up ℤ).Rel n₀ n₁
+  · ext ⟨p, q⟩ hpq
+    dsimp at h hpq
+    dsimp [totalShift₁XIso]
+    rw [ι_D₂_assoc, Linear.comp_units_smul, ι_totalDesc_assoc, ι_D₂,
+      ((shiftFunctor₁ C x).obj K).d₂_eq _ _ rfl _ (by dsimp; omega),
+      K.d₂_eq _ _ rfl _ (by dsimp; omega), smul_smul,
+      Linear.units_smul_comp, assoc, ι_totalDesc]
+    dsimp
+    congr 1
+    rw [add_comm p, Int.negOnePow_add, ← mul_assoc, Int.units_mul_self, one_mul]
+  · rw [D₂_shape _ _ _ _ h, zero_comp, D₂_shape, comp_zero, smul_zero]
+    intro h'
+    apply h
+    dsimp at h' ⊢
+    omega
+
+/-- The isomorphism `((shiftFunctor₁ C x).obj K).total (up ℤ) ≅ (K.total (up ℤ))⟦x⟧`
+expressing the compatibility of the total complex with the shift on the first indices.
+This isomorphism does not involve signs. -/
+noncomputable def totalShift₁Iso :
+    ((shiftFunctor₁ C x).obj K).total (up ℤ) ≅ (K.total (up ℤ))⟦x⟧ :=
+  HomologicalComplex.Hom.isoOfComponents (fun n => K.totalShift₁XIso x n (n + x) rfl)
+    (fun n n' _ => by
+      dsimp
+      simp only [Preadditive.add_comp, Preadditive.comp_add, smul_add,
+        Linear.comp_units_smul, K.D₁_totalShift₁XIso_hom x n n' _ _ rfl rfl,
+        K.D₂_totalShift₁XIso_hom x n n' _ _ rfl rfl])
+
+end
+
+section
+
+variable (y : ℤ) [K.HasTotal (up ℤ)] [((shiftFunctor₂ C y).obj K).HasTotal (up ℤ)]
+
+attribute [local simp] smul_smul
+
+/-- Auxiliary definition for `totalShift₂Iso`. -/
+noncomputable def totalShift₂XIso (n n' : ℤ) (h : n + y = n') :
+    (((shiftFunctor₂ C y).obj K).total (up ℤ)).X n ≅ (K.total (up ℤ)).X n' where
+  hom := totalDesc _ (fun p q hpq => (p * y).negOnePow • K.ιTotal (up ℤ) p (q + y) n'
+    (by dsimp at hpq ⊢; omega))
+  inv := totalDesc _ (fun p q hpq => (p * y).negOnePow •
+    (K.XXIsoOfEq rfl (Int.sub_add_cancel q y)).inv ≫
+      ((shiftFunctor₂ C y).obj K).ιTotal (up ℤ) p (q - y) n (by dsimp at hpq ⊢; omega))
+  hom_inv_id := by
+    ext p q h
+    dsimp
+    simp only [ι_totalDesc_assoc, Linear.units_smul_comp, ι_totalDesc, smul_smul,
+      Int.units_mul_self, one_smul, comp_id]
+    exact ((shiftFunctor₂ C y).obj K).XXIsoOfEq_inv_ιTotal _ rfl (by omega) _ _
+
+lemma D₁_totalShift₂XIso_hom (n₀ n₁ n₀' n₁' : ℤ) (h₀ : n₀ + y = n₀') (h₁ : n₁ + y = n₁') :
```

Comment:
Should this get `@[reassoc]`? Same for `D₂_totalShift₂XIso_hom` below and the two analogous lemmas above.

Reply:
Thanks! Yes, I have just added the reassoc attribute.